File: lucafontanili.designpatterns.java/src/main/java/com/lucafontanili/designpatterns/behavioral/visitor/VisitorClient.java
package com.lucafontanili.designpatterns.behavioral.visitor;

import java.io.IOException;

import com.lucafontanili.designpatterns.utilities.AbstractClient;
import com.lucafontanili.designpatterns.utilities.InvalidArgumentException;

public class VisitorClient extends AbstractClient {

    public static void main(String[] args) throws InvalidArgumentException, IOException {
	main(new VisitorClient());

    }

    @Override
    public void run() throws InvalidArgumentException, IOException {
	AbstractEquipment floppy = new FloppyDiskEquipment();
	floppy.setNetPrice(10);
	floppy.setDiscountPrice(7);
	FloppyDiskVisitor floppyDiskVisitor = new FloppyDiskVisitor();
	floppy.accept(floppyDiskVisitor);
	LOGGER.info(new StringBuilder(32).append("Total floppy price: ").append(floppyDiskVisitor.getTotal()));

	ChassisEquipment chassis = new ChassisEquipment();
	chassis.addComponent(new ComputerEquipment(), 8, 6, "Bus");
	chassis.addComponent(new ComputerEquipment(), 4, 2, "RAM");
	ChassisVisitor chassisVisitor = new ChassisVisitor();
	chassis.accept(chassisVisitor);
	LOGGER.info(new StringBuilder(32).append("Inventory: ").append(chassisVisitor.getInventory()));

    }

}

File: lucafontanili.designpatterns.python/src/behavioral/visitor/VisitorClient.py
'''
Created on Nov 29, 2016

@author: Admin
'''

from Visitor import *
from Equipment import *

def main():
    floppy = FloppyDiskEquipment(10, 7, 'floppy')
    floppy_disk_visitor = FloppyDiskVisitor()
    floppy.accept(floppy_disk_visitor)
    print('Total floppy price: ' + str(floppy_disk_visitor.total()))
    
    chassis = ChassisEquipment()
    chassis.add_component(8, 6, 'Bus')
    chassis.add_component(4, 2, 'RAM')
    chassis_visitor = ChassisVisitor()
    chassis.accept(chassis_visitor)
    print('Inventory:')
    [component.to_string() for component in chassis_visitor.inventory()]

if __name__ == "__main__":
    main()